### PR TITLE
Feature/test exception messages

### PR DESCRIPTION
**Pull request by fdlk**
Changes:

* cherrypicks Connor's high sierra fix to the global exception handling branch cause else I cannot run molgenis :)
* Add module `molgenis-i18n` which sits somewhere at the bottom of the module tree and doesn't depend on data service. Move interfaces and implementations from `molgenis-data-i18n` to there whenever possible.
* Introduce MolgenisLocaleResolver bean. 
    * Means that you can call LocaleContextHolder.getLocale() to retrieve locale to use to format messages.
* Introduce MessageSource bean. 
    * MessageSourceHolder with static getter for non-beans that do need access to it.
    * Implementation in the webapp uses LocalizationService to retrieve message formats from the dataservice.
    * Implementation for unit tests uses TestAllPropertiesMessageSource which is properties based
* LanguageServiceImpl -> LanguageService static methods. Deprecate those that can be replaced with calls to the methods above.
* Change bootstrapping a bit
    * Populator uses single AllPropertiesMessageSource, fills it with namespaces from MessagePropertySource beans (creation of these is unchanged, but they are now just a wrapper around the name of the bundle and don't actually retrieve messages)

#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- (n.a.) User documentation updated
- (n.a.) (If you have changed REST API interface) view-swagger.ftl updated
- (n.a.) Test plan template updated
- [x] Clean commits

### CODE REVIEW

**Comment by tommydeboer** [comment]
If we add exception tests to test the localized messages, wouldn't that mean that the people that change the messages also need to understand how to fix all the tests?

File: molgenis-core-ui/pom.xml [hunk 1]
```diff
@@ -136,7 +136,7 @@
                                 <location>org.molgenis.ui.admin.user</location>
                                 <location>org.molgenis.ui.admin.usermanager</location>
                             </locations>
-                            <swaggerDirectory>${build.directory}/generated-resources/swagger</swaggerDirectory>
+                            <swaggerDirectory>${project.build.directory}/generated-resources/swagger</swaggerDirectory>
```
[the remark this answers is not in the document]
Comment (fdlk) [reply molgenis-core-ui/pom.xml]: Something that maven complains about every time the build runs. I thought I'd sneak it in here.

**Comment by fdlk** [comment]
@tommydeboer I think it means that non-tech-savvy people can update and create resources for the messages based on the existing ones. But when they want to turn them into the molgenis defaults they give them to a developer who updates the tests and in the meanwhile also tests that they're correct when it comes to the fancier message format strings. (Using '' for example is hard to explain.)

File: molgenis-data-validation/src/main/java/org/molgenis/data/validation/ValidationException.java [hunk 2]
```diff
@@ -39,18 +38,16 @@ public String getMessage()
 	@Override
 	public String getLocalizedMessage()
 	{
-		String localizedViolationsMessage = validationMessages.stream()
-															  .map(ValidationMessage::getLocalizedMessage)
-															  .collect(joining("\n"));
-		return getLanguageService().map(
-				languageService -> languageService.getString(ERROR_CODE) + ": " + localizedViolationsMessage)
-								   .orElseGet(super::getLocalizedMessage);
+		//TODO: put the validation messages as argument in the message format
```
[the remark this answers is not in the document]
Comment (fdlk) [reply molgenis-data-validation/src/main/java/org/molgenis/data/validation/ValidationException.java]: Regards how we handle ValidationExceptions. 
This PR doesn't change what you can do but I think ideally we'd look more into how spring-validation does it and steal some of their ideas.

**Comment by tommydeboer** [comment]
I'm getting a stack overflow during bootstrapping:

```
com.google.common.util.concurrent.ExecutionError: java.lang.StackOverflowError
	at com.github.benmanes.caffeine.guava.CaffeinatedGuavaLoadingCache.getUnchecked(CaffeinatedGuavaLoadingCache.java:83)
	at org.molgenis.data.cache.l2.L2Cache.get(L2Cache.java:91)
	at org.molgenis.data.cache.l2.L2CacheRepositoryDecorator.findOneById(L2CacheRepositoryDecorator.java:66)
	at org.molgenis.data.cache.l1.L1CacheRepositoryDecorator.findOneById(L1CacheRepositoryDecorator.java:75)
	at org.molgenis.data.AbstractRepositoryDecorator.findOneById(AbstractRepositoryDecorator.java:107)
	at org.molgenis.data.AbstractRepositoryDecorator.findOneById(AbstractRepositoryDecorator.java:107)
	at org.molgenis.data.AbstractRepositoryDecorator.findOneById(AbstractRepositoryDecorator.java:107)
	at org.molgenis.data.AbstractRepositoryDecorator.findOneById(AbstractRepositoryDecorator.java:107)
	at org.molgenis.data.EntityReferenceResolverDecorator.findOneById(EntityReferenceResolverDecorator.java:63)
	at org.molgenis.data.AbstractRepositoryDecorator.findOneById(AbstractRepositoryDecorator.java:107)
	at org.molgenis.data.AbstractRepositoryDecorator.findOneById(AbstractRepositoryDecorator.java:107)
	at org.molgenis.data.AbstractRepositoryDecorator.findOneById(AbstractRepositoryDecorator.java:107)
	at org.molgenis.data.RepositorySecurityDecorator.findOneById(RepositorySecurityDecorator.java:79)
	at org.molgenis.data.transaction.TransactionalRepositoryDecorator.lambda$findOneById$6(TransactionalRepositoryDecorator.java:82)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at org.molgenis.data.transaction.TransactionalRepositoryDecorator.findOneById(TransactionalRepositoryDecorator.java:82)
	at org.molgenis.data.AbstractRepositoryDecorator.findOneById(AbstractRepositoryDecorator.java:107)
	at org.molgenis.data.support.DataServiceImpl.findOneById(DataServiceImpl.java:75)
	at org.molgenis.data.support.DataServiceImpl$$FastClassBySpringCGLIB$$2e5f77dc.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:669)
	at org.molgenis.data.support.DataServiceImpl$$EnhancerBySpringCGLIB$$796b929.findOneById(<generated>)
	at org.molgenis.data.settings.DefaultSettingsEntity.lambda$getEntity$3(DefaultSettingsEntity.java:227)
	at org.molgenis.security.core.runas.RunAsSystemAspect.runAsSystem(RunAsSystemAspect.java:44)
	at org.molgenis.data.settings.DefaultSettingsEntity.getEntity(DefaultSettingsEntity.java:225)
	at org.molgenis.data.settings.DefaultSettingsEntity.getString(DefaultSettingsEntity.java:82)
	at org.molgenis.ui.settings.AppDbSettings.getLanguageCode(AppDbSettings.java:249)
	at org.molgenis.data.i18n.LocalizationConfig.lambda$messageSource$1(LocalizationConfig.java:54)
	at org.molgenis.i18n.LocalizationMessageSource.resolveCodeWithoutArguments(LocalizationMessageSource.java:84)
	at org.springframework.context.support.AbstractMessageSource.getMessageInternal(AbstractMessageSource.java:209)
	at org.springframework.context.support.AbstractMessageSource.getMessage(AbstractMessageSource.java:136)
	at org.molgenis.data.ErrorCodedDataAccessException.getLocalizedMessage(ErrorCodedDataAccessException.java:38)
	at java.lang.Throwable.toString(Throwable.java:480)
	at java.lang.Throwable.<init>(Throwable.java:311)
	at java.lang.Exception.<init>(Exception.java:102)
	at java.lang.RuntimeException.<init>(RuntimeException.java:96)
	at com.google.common.util.concurrent.UncheckedExecutionException.<init>(UncheckedExecutionException.java:61)
	at com.github.benmanes.caffeine.guava.CaffeinatedGuavaLoadingCache.getUnchecked(CaffeinatedGuavaLoadingCache.java:81)
	at org.molgenis.data.cache.l2.L2Cache.get(L2Cache.java:91)
	at org.molgenis.data.cache.l2.L2CacheRepositoryDecorator.findOneById(L2CacheRepositoryDecorator.java:66)
	at org.molgenis.data.cache.l1.L1CacheRepositoryDecorator.findOneById(L1CacheRepositoryDecorator.java:75)
etc.
```

Will try with an empty database now.

**Comment by tommydeboer** [comment]
When going to:
[http://localhost:8080/menu/main/dataexplorer?entity=aap](http://localhost:8080/menu/main/dataexplorer?entity=aap)

I get an empty error message:
```Error! (D01)```

**Comment by tommydeboer** [comment]
Some elements in the UI aren't localized:

#dataexplorer_data_data_item_filters#
#dataexplorer_aggregates_title#

File: molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/core/exception/AnnotationException.java [hunk 3]
```diff
@@ -59,25 +56,14 @@ public String getMessage()
...
-					StringUtils.join(requiredAttributeNames, ","), annotatorName, cause);
-		}).orElse(super.getLocalizedMessage());
-	}
-
 	@Override
 	protected Object[] getLocalizedMessageArguments()
 	{
-		throw new UnsupportedOperationException();
+		String languageCode = LocaleContextHolder.getLocale().getLanguage();
+		// Annotation with annotator:''{3}'' failed for entity:''{0,label}'' on row: {1}, required attributes for this annotator:''{2}''
```
[the remark this answers is not in the document]
Comment (fdlk) [reply molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/core/exception/AnnotationException.java]: yes, removing!

File: molgenis-data-security/src/test/java/org/molgenis/data/security/exception/EntityTypePermissionDeniedExceptionTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import org.mockito.Mock;
+import org.molgenis.data.meta.model.EntityType;
+import org.molgenis.security.core.Permission;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.when;
+import static org.testng.Assert.assertEquals;
+
+public class EntityTypePermissionDeniedExceptionTest extends org.molgenis.i18n.test.exception.ExceptionMessageTest
```
[the remark this answers is not in the document]
Comment (fdlk) [reply molgenis-data-security/src/test/java/org/molgenis/data/security/exception/EntityTypePermissionDeniedExceptionTest.java]: done!

**Pull request merged**